include/deal.II/base/quadrature_point_data.h

```diff
@@ -168,6 +168,12 @@ class CellDataStorage : public Subscriptor
   std::vector<std::shared_ptr<const T>>
   get_data(const CellIteratorType &cell) const;
 
+  /**
+   * Queries whether a data belonging to @p cell exists.
+   */
+  bool
+  stores_data(const CellIteratorType &cell) const;
```

Comment:
It's a shame that we didn't anticipate the fact that some cells might not have data associated with them, or we could have let `get_data()` return a `boost::optional` object.

Reply:
I also think changing the output of `get_data()` to give additional info on whether a data exists would have been a much better design. We discussed it in #8506 and decided to go with this design (adding member function `stores_data()` instead).

I prefer writing
```c++
auto qpd = data_storage(cell);
if (!qpd)
  return;
```
instead of the current expression
```c++
if (!data_storage->stores_data(cell))
    return;
auto qpd = data_storage->get_data(cell);
```
because the latter requires two map lookups instead of just one lookup.
However, backward compatibility limits us a bit here because it forces us to choose the second design.

Reply:
Or do `if (auto data = data_storage->get_data) { ... }` :-)

Like I mentioned elsewhere, we always have the option of deprecating `get_data()` and invent a new function that has the right interface.

Reply:
I'm happy if you want to follow through on @bangerth's suggestion if it seems like the better solution to the majority of people. @bangerth Maybe you wouldn't mind reviewing this again once its been updated -- I'm still short on time ATM.

Reply:
@jppelteret -- yes, gladly!
@rezarastak -- let me know once you want me to take another look!